### PR TITLE
refactor(www): Split up gatsby-node

### PR DESCRIPTION
**Pull request by dvrylc**
## Description

Split up `gatsby-node` into logical components. 

Pending tasks/questions: 
- [x] Switch `createPages` to async once #21210 merges
- [x] Should we have a `contributors.js` too? 

## Related Issues

Addresses #21209.

### CODE REVIEW

File: www/src/utils/node/creators.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
+const Promise = require(`bluebird`)
+const path = require(`path`)
+const slash = require(`slash`)
+const slugify = require(`slugify`)
+
+exports.createPages = ({ graphql, actions }) => {
+  const { createPage } = actions
+
+  return new Promise((resolve, reject) => {
```
Comment (tesseralis) [inline www/src/utils/node/creators.js]: Could we use [async/await syntax](https://github.com/gatsbyjs/gatsby/pull/21210/files#diff-fb043578f52339bcb2a2d2b4df1203c8R112) for these? While we're refactoring, we might as well pull everything up to date with modern syntax :)

File: www/src/utils/node/creators.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const creatorPageTemplate = path.resolve(
+      `src/templates/template-creator-details.js`
+    )
+
+    graphql(`
+      query {
+        allCreatorsYaml {
+          edges {
```
Comment (tesseralis) [inline www/src/utils/node/creators.js]: We can actually just use:

```
allCreatorsYaml {
  nodes {
    fields {
      slug
    }
  }
}

File: www/src/utils/node/docs.js [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+          path: `/blog/tags/${tagSlug}/`,
+          component: tagTemplate,
+          context: {
+            tags,
+          },
+        })
+      })
+
+      // Create docs pages.
+      const docPages = result.data.allMdx.edges
```
Comment (tesseralis) [inline www/src/utils/node/docs.js]: As a note, there's another PR I made to factor out the prev-and-next logic: #20978

File: www/gatsby-node.js [hunk 4]
```diff
@@ -1,76 +1,27 @@
...
-const docSlugFromPath = parsedFilePath => {
-  if (parsedFilePath.name !== `index` && parsedFilePath.dir !== ``) {
-    return `/${parsedFilePath.dir}/${parsedFilePath.name}/`
-  } else if (parsedFilePath.dir === ``) {
-    return `/${parsedFilePath.name}/`
-  } else {
-    return `/${parsedFilePath.dir}/`
-  }
-}
+let ecosystemFeaturedItems
```
Comment (tesseralis) [inline www/gatsby-node.js]: Note: this PR will change things here if it merges first: #21200

File: www/src/utils/node/creators.js (hunk 1 above)
Comment (tesseralis) [reply www/src/utils/node/creators.js]: Just read your TODO item -- we don't have to wait. You can just convert it to async here, and we can close out my PR.

**Comment by tesseralis** [comment]
For `contributior` pages -- I would actually put them in `docs.js`. The reasoning is that the authors are associated with blog posts, which are in `docs.js`. Ideally we'd split up docs/tutorial/contributing from blogs but that's a more intricate refactor and can wait for later.

**Comment by dvrylc** [comment]
I've done the following: 

-  Use `async`/`await` instead of promises for `createPages`
- Use `nodes` in graphql queries
- Grouped contributor pages with docs

Regarding #20978 and #21200, I think we should leave those as separate PRs since they aren't really related to splitting up gatsby-node, and this PR is getting pretty large. What do you think @tesseralis?

**Comment by tesseralis** [comment]
@dvrylc agreed that they should be separate. I just linked them for reference in case they got merged first and caused a conflict.

File: www/gatsby-node.js [hunk 5]
```diff
@@ -85,661 +35,47 @@ exports.createPages = ({ graphql, actions, reporter }) => {
...
-    }
-    slug = `/creators/${validTypes[node.type]}/${slugify(node.name, {
-      lower: true,
-    })}`
-    createNodeField({ node, name: `slug`, value: slug })
-  }
-  // end Creator pages
+exports.onCreateNode = helpers => {
+  sections.forEach(section => section.onCreateNode(helpers))
   return null
```
Comment (tesseralis) [inline www/gatsby-node.js]:
```suggestion
```

File: www/gatsby-node.js [hunk 6]
```diff
@@ -85,661 +35,47 @@ exports.createPages = ({ graphql, actions, reporter }) => {
...
-          path,
-          component: slash(featureComparisonPageTemplate),
-          context: {
-            options,
-            featureType,
-          },
-        })
-      }
+  // Create feature comparison pages
+  const jamstackPages = generateComparisonPageSet(`jamstack`)
```
Comment (tesseralis) [inline www/gatsby-node.js]: features probably goes in its own file anyways, but we can do this in another PR.

File: www/src/utils/node/creators.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+const Promise = require(`bluebird`)
```
Comment (tesseralis) [inline www/src/utils/node/creators.js]: could you go over and make sure to remove the unused imports from these files?

**Comment by dvrylc** [comment]
Thanks for the comments! I've cleaned up the code, rebased and separated features into its own file (`features.js`). 

Should be good to go 🚀

**Comment by pieh** [comment]
This PR seems to be 1 step away from making those actual gatsby themes and with that we wouldn't need the "sections" pattern / importing "gatsby-node".

Instead we could move `www/src/utils/node/creators.js` to `www/plugins/gatsby-dot-org-creators/gatsby-node.js`, add empty package.json (well it does need to have `{}`) and empty index.js and add `gatsby-dot-org-creators` to `www/gatsby-config.js` and it would work (unless there is some ordering that actually matters).

**Comment by tesseralis** [comment]
@pieh that's the point. I've tried moving around stuff to themes a couple of times and there's always been some unforeseen problem (things are connected in a weird way, etc.) We could just move the gatsby-node files to plugins but I was thinking of moving everything related to a theme at once when we make the themes -- for creators, that would be all related components, the `gatsby-config` with all dependencies related to it, etc. What do you think?

**Comment by pieh** [comment]
> @pieh that's the point. I've tried moving around stuff to themes a couple of times and there's always been some unforeseen problem (things are connected in a weird way, etc.) We could just move the gatsby-node files to plugins but I was thinking of moving everything related to a theme at once when we make the themes -- for creators, that would be all related components, the `gatsby-config` with all dependencies related to it, etc. What do you think?

Up to you - I have no preference here.

**Comment by tesseralis** [comment]
@pieh okay--let's just merge this as-is then.

**Comment by pieh** [comment]
There is conflict needed to be resolved (probably after merging the JSdoc/MDX PR)

**Comment by tesseralis** [comment]
@dvrylc I can handle this conflict, since it was my PRs that caused it.

File: www/src/utils/node/docs.js [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+  const blogListTemplate = path.resolve(`src/templates/template-blog-list.js`)
+  const tagTemplate = path.resolve(`src/templates/tags.js`)
+  const contributorPageTemplate = path.resolve(
+    `src/templates/template-contributor-page.js`
+  )
+  const localPackageTemplate = path.resolve(
+    `src/templates/template-docs-local-packages.js`
+  )
+
+  const result = await graphql(`
```
Comment (tesseralis) [inline www/src/utils/node/docs.js]: just a nit, but I prefer

```
const { data, errors } = await graphql(`...`)
if (errors) throw errors

data.nodes.forEach(node => {
 // do something
})

Comment (dvrylc) [reply www/src/utils/node/docs.js]: Nice, updated the PR!

**Comment by dvrylc** [comment]
> @dvrylc I can handle this conflict, since it was my PRs that caused it.

Only saw this after going through the long rebase 😓 but thanks for resolving it!
